File: src/cli-display.ts
/**
 * CLI display utilities — file classification, summary table formatting,
 * and pipeline result display.
 *
 * Migrated from Python `__main__.py` display functions.
 *
 * @module
 */
import type { PipelineResult } from "./models.js";
import { isKnowledgePackFile } from "./cli-kp-detect.js";

export { isKnowledgePackFile } from "./cli-kp-detect.js";

/** Category counts for the summary table. */
export interface ComponentCounts {
  readonly rules: number;
  readonly skills: number;
  readonly knowledgePacks: number;
  readonly agents: number;
  readonly hooks: number;
  readonly settings: number;
  readonly readme: number;
  readonly github: number;
  readonly codex: number;
}

const SEPARATOR_CHAR = "\u2500";
const HEADER_LABEL = "Component";
const HEADER_COUNT = "Files";
const TOTAL_LABEL = "Total";

/** Ordered label-to-key mapping for the summary table. */
const CATEGORY_LABELS: ReadonlyArray<readonly [string, keyof ComponentCounts]> = [
  ["Rules (.claude)", "rules"],
  ["Skills (.claude)", "skills"],
  ["Knowledge Packs (.claude)", "knowledgePacks"],
  ["Agents (.claude)", "agents"],
  ["Hooks (.claude)", "hooks"],
  ["Settings (.claude)", "settings"],
  ["README", "readme"],
  ["GitHub", "github"],
  ["Codex", "codex"],
];

/** Split a file path into its directory and filename segments. */
function splitPathSegments(filePath: string): readonly string[] {
  return filePath.split(/[\\/]/);
}

/**
 * Classify a single file path into a component category.
 *
 * @param filePath - Path to classify.
 * @param segments - Pre-split path segments.
 * @param kpCache - Cache for knowledge pack detection.
 * @returns The matching category key, or undefined if no match.
 */
function classifySingleFile(
  filePath: string,
  segments: readonly string[],
  kpCache: Map<string, boolean>,
): keyof ComponentCounts | undefined {
  const fileName = segments[segments.length - 1] ?? "";
  if (segments.includes(".agents")) return "codex";
  if (segments.includes(".codex")) return "codex";
  if (segments.includes(".github")) return "github";
  if (fileName.includes("README")) return "readme";
  if (fileName.includes("settings")) return "settings";
  if (segments.includes("hooks")) return "hooks";
  if (segments.includes("agents")) return "agents";
  if (segments.includes("skills")) {
    // In dry-run mode, files may not exist on disk (temp dir deleted).
    // isKnowledgePackFile returns false for missing files — matches Python behavior.
    return isKnowledgePackFile(filePath, kpCache)
      ? "knowledgePacks"
      : "skills";
  }
  if (segments.includes("rules")) return "rules";
  if (fileName === "AGENTS.md") return "codex";
  return undefined;
}

/**
 * Classify generated file paths into component categories.
 *
 * Categorizes by path segments (github, rules, skills, agents, hooks)
 * and file name patterns (README, settings). GitHub takes priority
 * over other categories. For skills, reads SKILL.md to detect
 * knowledge packs.
 *
 * @param filePaths - Array of generated file paths.
 * @returns Counts for each component category.
 */
export function classifyFiles(
  filePaths: readonly string[],
): ComponentCounts {
  const counts: Record<keyof ComponentCounts, number> = {
    rules: 0,
    skills: 0,
    knowledgePacks: 0,
    agents: 0,
    hooks: 0,
    settings: 0,
    readme: 0,
    github: 0,
    codex: 0,
  };

  const kpCache = new Map<string, boolean>();
  for (const filePath of filePaths) {
    const segments = splitPathSegments(filePath);
    const category = classifySingleFile(filePath, segments, kpCache);
    if (category !== undefined) {
      counts[category]++;
    }
  }

  return counts;
}

/** Compute the total count from all categories. */
function computeTotal(counts: ComponentCounts): number {
  return CATEGORY_LABELS.reduce(
    (sum, [, key]) => sum + counts[key],
    0,
  );
}

/** Compute the label column width for alignment. */
function computeLabelWidth(counts: ComponentCounts): number {
  let maxWidth = HEADER_LABEL.length;
  for (const [label, key] of CATEGORY_LABELS) {
    if (counts[key] > 0 && label.length > maxWidth) {
      maxWidth = label.length;
    }
  }
  if (TOTAL_LABEL.length > maxWidth) {
    maxWidth = TOTAL_LABEL.length;
  }
  return maxWidth;
}

/** Build the data rows (non-zero categories) for the summary table. */
function buildDataRows(
  counts: ComponentCounts,
  labelWidth: number,
  countWidth: number,
): string[] {
  const rows: string[] = [];
  for (const [label, key] of CATEGORY_LABELS) {
    if (counts[key] > 0) {
      const countStr = String(counts[key]).padStart(countWidth);
      rows.push(`  ${label.padEnd(labelWidth)}  ${countStr}`);
    }
  }
  return rows;
}

/**
 * Format the summary table string for terminal output.
 *
 * Only displays categories with count > 0. Includes header,
 * separator lines using U+2500 (box-drawing horizontal), and total row.
 *
 * @param counts - Component counts to display.
 * @returns Formatted table string ready for console output.
 */
export function formatSummaryTable(
  counts: ComponentCounts,
): string {
  const total = computeTotal(counts);
  const labelWidth = computeLabelWidth(counts);
  const countWidth = HEADER_COUNT.length;
  const sep = SEPARATOR_CHAR.repeat(labelWidth);
  const cSep = SEPARATOR_CHAR.repeat(countWidth);
  const header = `  ${HEADER_LABEL.padEnd(labelWidth)}  ${HEADER_COUNT}`;
  const divider = `  ${sep}  ${cSep}`;
  const dataRows = buildDataRows(counts, labelWidth, countWidth);
  const totalRow = `  ${TOTAL_LABEL.padEnd(labelWidth)}  ${String(total).padStart(countWidth)}`;

  return [header, divider, ...dataRows, divider, totalRow].join("\n");
}

/**
 * Display the full pipeline result: success line, summary table,
 * output directory, and warnings.
 *
 * @param result - The pipeline result to display.
 * @throws Error if `result.success` is false.
 */
export function displayResult(result: PipelineResult): void {
  if (!result.success) {
    throw new Error("Pipeline did not succeed");
  }

  console.log(`Pipeline: Success (${result.durationMs}ms)`);
  console.log();

  const counts = classifyFiles(result.filesGenerated);
  const table = formatSummaryTable(counts);
  console.log(table);

  console.log();
  console.log(`Output: ${result.outputDir}`);

  for (const warning of result.warnings) {
    console.log(`Warning: ${warning}`);
  }
}

File: java/src/main/java/dev/iadev/cli/CliDisplay.java
package dev.iadev.cli;

import dev.iadev.domain.model.PipelineResult;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Formats pipeline results for terminal display.
 *
 * <p>File path categorization is delegated to
 * {@link FileCategorizer}.</p>
 *
 * @see PipelineResult
 * @see FileCategorizer
 */
public final class CliDisplay {

    private static final String SEPARATOR_CHAR = "\u2500";
    private static final String HEADER_LABEL = "Category";
    private static final String HEADER_COUNT = "Count";
    private static final String TOTAL_LABEL = "Total";
    private static final int MIN_LABEL_WIDTH = 20;

    private CliDisplay() {
        // utility class
    }

    /**
     * Classifies file paths into categories.
     *
     * @param paths the list of file paths to classify
     * @return ordered map of category to paths
     */
    public static Map<String, List<String>> classifyFiles(
            List<String> paths) {
        Map<String, List<String>> classified =
                new LinkedHashMap<>();

        for (String path : paths) {
            String normalized =
                    FileCategorizer.normalizePath(path);
            String category =
                    FileCategorizer.categorize(normalized);
            classified.computeIfAbsent(
                    category, k -> new ArrayList<>())
                    .add(path);
        }

        return classified;
    }

    /**
     * Formats a summary table from the classified file map.
     *
     * @param classified the classified file map
     * @return formatted table string
     */
    public static String formatSummaryTable(
            Map<String, List<String>> classified) {
        int total = classified.values().stream()
                .mapToInt(List::size).sum();
        int labelWidth = computeLabelWidth(classified);
        int countWidth = HEADER_COUNT.length();

        StringBuilder sb = new StringBuilder();
        appendHeader(sb, labelWidth, countWidth);
        appendCategoryRows(
                sb, classified, labelWidth, countWidth);
        appendFooter(
                sb, total, labelWidth, countWidth);

        return sb.toString();
    }

    private static void appendHeader(
            StringBuilder sb,
            int labelWidth, int countWidth) {
        String sep = SEPARATOR_CHAR.repeat(labelWidth);
        String cSep = SEPARATOR_CHAR.repeat(countWidth);
        sb.append("  ")
                .append(pad(HEADER_LABEL, labelWidth))
                .append("  ").append(HEADER_COUNT)
                .append('\n');
        sb.append("  ").append(sep)
                .append("  ").append(cSep).append('\n');
    }

    private static void appendCategoryRows(
            StringBuilder sb,
            Map<String, List<String>> classified,
            int labelWidth, int countWidth) {
        for (Map.Entry<String, List<String>> entry
                : classified.entrySet()) {
            int count = entry.getValue().size();
            if (count > 0) {
                sb.append("  ")
                        .append(pad(entry.getKey(),
                                labelWidth))
                        .append("  ")
                        .append(padLeft(
                                String.valueOf(count),
                                countWidth))
                        .append('\n');
            }
        }
    }

    private static void appendFooter(
            StringBuilder sb, int total,
            int labelWidth, int countWidth) {
        String sep = SEPARATOR_CHAR.repeat(labelWidth);
        String cSep = SEPARATOR_CHAR.repeat(countWidth);
        sb.append("  ").append(sep)
                .append("  ").append(cSep).append('\n');
        sb.append("  ")
                .append(pad(TOTAL_LABEL, labelWidth))
                .append("  ")
                .append(padLeft(
                        String.valueOf(total),
                        countWidth));
    }

    /**
     * Formats the complete pipeline result for display.
     *
     * @param result      the pipeline result
     * @param displayMode the display mode
     * @return formatted result string
     */
    public static String formatResult(
            PipelineResult result,
            DisplayMode displayMode) {
        StringBuilder sb = new StringBuilder();
        appendPipelineHeader(sb, result, displayMode);
        appendSummarySection(sb, result);
        appendDryRunDetails(sb, result, displayMode);
        appendWarnings(sb, result);
        return sb.toString();
    }

    private static void appendPipelineHeader(
            StringBuilder sb, PipelineResult result,
            DisplayMode displayMode) {
        if (displayMode.isDryRun()) {
            sb.append("[DRY RUN] ");
        }
        sb.append("Pipeline: Success (")
                .append(result.durationMs())
                .append("ms)\n\n");
    }

    private static void appendSummarySection(
            StringBuilder sb, PipelineResult result) {
        Map<String, List<String>> classified =
                classifyFiles(result.filesGenerated());
        sb.append(formatSummaryTable(classified));
        sb.append("\n\n");
        sb.append("Output: ").append(result.outputDir());
    }

    private static void appendDryRunDetails(
            StringBuilder sb, PipelineResult result,
            DisplayMode displayMode) {
        if (displayMode.isDryRun()
                && !result.filesGenerated().isEmpty()) {
            sb.append(
                    "\n\nFiles that would be generated:\n");
            for (String file : result.filesGenerated()) {
                sb.append("  ").append(file).append('\n');
            }
        }
    }

    private static void appendWarnings(
            StringBuilder sb, PipelineResult result) {
        for (String warning : result.warnings()) {
            sb.append("\nWarning: ").append(warning);
        }
    }

    /**
     * Formats verbose output for an assembler execution.
     *
     * @param assemblerName the assembler name
     * @param fileCount     number of files generated
     * @param durationMs    execution time in milliseconds
     * @return formatted verbose line
     */
    public static String formatAssemblerVerbose(
            String assemblerName,
            int fileCount,
            long durationMs) {
        return "%s completed in %dms (%d files)"
                .formatted(assemblerName,
                        durationMs, fileCount);
    }

    private static int computeLabelWidth(
            Map<String, List<String>> classified) {
        int maxWidth = Math.max(
                HEADER_LABEL.length(), MIN_LABEL_WIDTH);
        for (String label : classified.keySet()) {
            if (label.length() > maxWidth) {
                maxWidth = label.length();
            }
        }
        if (TOTAL_LABEL.length() > maxWidth) {
            maxWidth = TOTAL_LABEL.length();
        }
        return maxWidth;
    }

    private static String pad(String s, int width) {
        if (s.length() >= width) {
            return s;
        }
        return s + " ".repeat(width - s.length());
    }

    private static String padLeft(String s, int width) {
        if (s.length() >= width) {
            return s;
        }
        return " ".repeat(width - s.length()) + s;
    }
}
